scripts/artifacts-building/user_rcbops_artifacts_building.yml

```diff
@@ -31,3 +31,66 @@ openstack_repo_url: "{{ rpco_mirror_base_url }}"
 # default variant.
 lxc_image_cache_server: images.linuxcontainers.org
 
+# When we build python/container artifacts we pull the default LXC cache
+# variant from LXC upstream which may have packages installed which are more
+# recent than those we have available in our artifacted apt repository. Here
+# we figure out which packages are installed, but not available in a
+# configured source, then we downgrade them to the latest available in the
+# configured sources.
+#
+# By way of reference, typically the output of 'apt list --installed' will
+# produce output such as the following:
+#
+# adduser/trusty,now 3.113+nmu3ubuntu3 all [installed]
+# ca-certificates/r14.0.0rc1-trusty,now 20160104ubuntu0.14.04.1 all [installed]
+# libc6/now 2.19-0ubuntu6.11 amd64 [installed,local]
+# libglib2.0-0/r14.0.0rc1-trusty,now 2.40.2-0ubuntu1 amd64 [installed,automatic]
+#
+# The format of the output is the following:
+# <package name>/<list of dist sources> <version> <arch> <flags>
+#
+# If a package is flagged as 'installed' and 'local' then it indicates that
+# the package has been installed from an apt source which is not currently
+# configured. These are the packages which need to be downgraded to a version
+# in an available apt source.
+#
+lxc_cache_prep_pre_commands: |
+    # If there is a configured resolver, save it.
+    if [ -a /etc/resolv.conf ]; then
+      mv /etc/resolv.conf /etc/resolv.conf.org
+    fi
+    # Use the LXC host's dnsmasq service as a resolver
+    echo "nameserver {{ lxc_net_address | default('10.0.3.1') }}" > /etc/resolv.conf
+    # Add the host's repository keys, including the RPC-O keys
+    apt-key add /root/repo.keys
+    # Update the apt cache
+    export DEBIAN_FRONTEND=noninteractive
+    apt-get update
+    # Check whether there are any installed packages which
+    # are not available in a configured source.
+    if apt list --installed 2>/dev/null | egrep '\[.*local.*\]'; then
```

Comment:
nit: to avoid querying twice we could

list_local_installed_pkg = $(apt list --installed 2>/dev/null | egrep '\[.*local.*\]')
if -z list_local_installed_pkg (or something like that); then
    pkg_downgrade_list = $( echo $list_local_installed_pkg | cut -d "/" -f 1  ) ...